docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1063 @@
+the `Tenant` tasks. We will create a new per-tenant metadata file that records
+the attachment generation number, and will also be used in future work for
+implementing fast migration, where the tenant has more states than simple
+attached & detached.
+
+#### Re-attachment on startup
+
+As it currently does, the pageserver may use local disk state to identify which tenants/timelines
+were attached before its most recent restart. However, it is now necessary
+to call out to the control plane to advance the generation number before
+doing any remote I/O for the attachment: we call this process **re-attaching**
+on startup.
+
+Advancing the generation number is necessary to avoid the pageserver assuming
+that it is running in an environment where a previous instance of "itself" must
+have been fenced before this instance started: e.g. if running in an environment
+with containers/VMs using shared block storage where some stale container/VM might
+still think it has the same tenants attached that we do.
+
+To avoid sending O(tenants) requests to the control plane, in practice
+we would send a batched request to re-attach many tenants at once.
+
+Re-attaching has two possible outcomes for each tenant:
+
+- Increment the generation number, and return the incremented number: this
+  tenant may now go active with the new generation number.
+- Do not increment the generation number: this tenant may start, but may
+  not write to remote storage. This case is for nodes that the control
+  plane is intentionally leaving on a stale generation, for example when
+  doing a migration between pageservers, and the origin pageserver restarts
+  during the migration.
+
+#### Reconciliation
+
+On attachment, the `Tenant` may have some local disk content from previous
+activity involving this tenant (e.g. it used to be attached here), and this
+needs reconciling with remote storage. Note that "reconcile" in this context is
+used generically, and does not refer specifically to the logic currently
+in the `reconcile_with_remote` function.
+
+Because pageservers have access to the WAL, they don't strictly have to
+use the latest remote storage contents at all: if the WAL history still contains
+the data required to bring the pageserver's local state up to date with
+the tip of the WAL, the pageserver could recover from the WAL and then
+overwrite whatever is in remote storage.
+
+However, it is preferable to recover from remote storage rather than
+to replay WAL, for the following reasons:
+
+- Simplicity: recovering from remote storage gives identical behavior
+  whether attaching a tenant for the first time or re-attaching later.
+- Efficiency: ingesting the WAL has significant cost beyond the I/O
+  of reading it (compaction, image layer generation). We should avoid
+  doing this work twice if there is already remote content available.
+- Reduced load on safekeepers, which are a less scalable resource than S3.
+
+#### Finding the remote indices for timelines
```

Comment:
This section should come right after the `#### Attachment` section above.
There you explained how the API changes.
This section is the logical progression of "what happens inside the API handler".

Reply:
editorial: moving sections